### ISSUE TITLE
Confusing warnings regarding unused (meta-)variables

### ISSUE THREAD

**Issue by Gaspi** [Deducteam/lambdapi#312]
The following file (excuse the old syntax) typechecks and computes successively `a` and `f (λx, x)` as expected. However it generates three warnings.
```
A : Type.
a : A.

def f : (A -> A) -> A.
[X] f (x => X) --> a.

#EVAL f (x => a).
#EVAL f (x => x).
```
- `[bug.dk, 5:12-13] Pattern variable [&X] does not need to be named.`
However replacing the rewrite rule with `[] f (x => _) --> a.` yields a different output as both `EVAL` now compute `a`. It seems that the wildcard `_` corresponds in lambdapi to a *fully applied* fresh variable which is a reasonable choice but different from Dedukti 2.7 where it is used as an *unapplied* fresh variable. In this particular case, even though X doesn't occur on the right-hand side, it cannot be replaced with a wildcard.
- `[bug.dk, 5:7-8] Variable [x] could be replaced by [_].`
- `[bug.dk, 7:9-10] Variable [x] could be replaced by [_].`
However replacing `(x => X)` with `(_ => X)` or `(x => a)` with `(_ => a)` both yield a parsing error: `Unexpected token [=>].`

**Comment by rlepigre** [comment]
We could probably make it so that in the legacy syntax a wildcard in a LHS is replaced by a pattern variable with no environment instead of a wildcard. In a LHS in Lambdapi, the behaviour is the following:
 - `&X`: named pattern variable with empty environment,
 - `&X[x,y,z]` same with an explicit environment,
 - `&_`: unnamed pattern variable with empty environment (what you want here),
 - `&_[x,y,z]` same with an explicit environment,
 - `_` unnamed pattern variable with full environment (what you currently get).

If we could replace the argument `x` with a wildcard, then it would not be in the environment and so the current wildcard would work. I implemented that quickly but that is not satisfactory if we want to emulate Dedukti's behaviour.

Fixing the problem correctly is not trivial, but it should not be too hard either.

**Comment by Gaspi** [comment]
If we want to emulate Dedukti, then the legacy syntax should simply map `_` to  `&_` instead of `_`.

The other two issues probably only require to fix the parser to allow lambda over a special `_` unnamed variable.

I forgot to add that the three warnings are the expected behavior here.

**Comment by rlepigre** [comment]
Yeah sure, I completely agree.

**Comment by rlepigre** [comment]
@Gaspi do we agree that the correct file (triggering no warning) should be:
```
A : Type.
a : A.

def f : (A -> A) -> A.
[] f (_ => _) --> a.

#EVAL f (_ => a).
#EVAL f (x => x).
```

And is this input accepted by Dedukti? More generally, where are wildcards accepted by Dedukti?

**Comment by Gaspi** [comment]
I agree. And the second `#EVAL` should print `f (λx, x)`.
I just checked and this is Dedukti's behavior with this file.
Dedukti doesn't warn against needlessly named variables but this feature should remain in lambdapi.

Dedukti would also allow:
- `f (x => _ x)`  (applied unnamed variables). Because we only have one wildcard it was defaulted to "unapplied" and applications have to be explicit.
It's would be nice to also have a symbol for "fully applied" (maybe `*`) but this is not in Dedukti.
- `_ : A -> A` as an alias for `A -> A` (for some reason...)

But doesn't allow anonymous "arguments" such as `def g (_ : A) (x:A) := x.` which is a bit weird because the only way around is to name the unused argument.

**Comment by rlepigre** [comment]
OK, this is fixed by 7121ba8cb8350a7f0b604b1eb9284d96664990b0.
